### ISSUE TITLE
1.14 masters flapping until all are upgraded

### ISSUE THREAD

**Issue by cann0nf0dder** [kubernetes/kubernetes#81286]
**What happened**:
 Master nodes upgraded from 1.13.7 to 1.14.4 flapping by going to NotReady every 50 sec until all the masters are upgraded.

All the pods running fine, no errors in kubelet on the upgraded master.
Kubelet remains running when I see NotReady in the logs and via `watch kubectl get nodes`

Upgraded master01 kube-controller-manager log

```
kubectl logs kube-controller-manager-c8r001
I0812 10:10:04.619981       1 serving.go:319] Generated self-signed cert in-memory
W0812 10:10:05.075278       1 authentication.go:249] No authentication-kubeconfig provided in order to lookup client-ca-file in configmap/extension-apiserver-authentication in kube-system, so client certificate authentication won't work.
W0812 10:10:05.075340       1 authentication.go:252] No authentication-kubeconfig provided in order to lookup requestheader-client-ca-file in configmap/extension-apiserver-authentication in kube-system, so request-header client certificate authentication won't work.
W0812 10:10:05.075363       1 authorization.go:146] No authorization-kubeconfig provided, so SubjectAccessReview of authorization tokens won't work.
I0812 10:10:05.075424       1 controllermanager.go:155] Version: v1.14.4
I0812 10:10:05.075867       1 secure_serving.go:116] Serving securely on [::]:10257
I0812 10:10:05.076383       1 deprecated_insecure_serving.go:51] Serving insecurely on [::]:10252
I0812 10:10:05.076572       1 leaderelection.go:217] attempting to acquire leader lease  kube-system/kube-controller-manager... 
```

Not upgraded master02 kube-controller-manager log
```
I0812 10:59:12.190364       1 event.go:221] Event(v1.ObjectReference{Kind:"Node", Namespace:"", Name:"c8r001", UID:"061df7cf-9da0-11e9-833b-005056b0ebb3", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeNotReady' Node c8r001 status is now: NodeNotReady
W0812 10:59:12.878291       1 garbagecollector.go:647] failed to discover some groups: map[networking.k8s.io/v1beta1:the server could not find the requested resource scheduling.k8s.io/v1:the server could not find the requested resource coordination.k8s.io/v1:the server could not find the requested resource node.k8s.io/v1beta1:the server could not find the requested resource]
E0812 10:59:20.111767       1 memcache.go:135] couldn't get resource list for networking.k8s.io/v1beta1: the server could not find the requested resource
E0812 10:59:20.117019       1 memcache.go:135] couldn't get resource list for scheduling.k8s.io/v1: the server could not find the requested resource
E0812 10:59:20.197310       1 memcache.go:135] couldn't get resource list for coordination.k8s.io/v1: the server could not find the requested resource
E0812 10:59:20.297424       1 memcache.go:135] couldn't get resource list for node.k8s.io/v1beta1: the server could not find the requested resource
E0812 10:59:33.528454       1 resource_quota_controller.go:407] unable to retrieve the complete list of server APIs: coordination.k8s.io/v1: the server could not find the requested resource, networking.k8s.io/v1beta1: the server could not find the requested resource, node.k8s.io/v1beta1: the server could not find the requested resource, scheduling.k8s.io/v1: the server could not find the requested resource
W0812 10:59:45.279350       1 garbagecollector.go:647] failed to discover some groups: map[networking.k8s.io/v1beta1:the server could not find the requested resource scheduling.k8s.io/v1:the server could not find the requested resource coordination.k8s.io/v1:the server could not find the requested resource node.k8s.io/v1beta1:the server could not find the requested resource]
E0812 10:59:50.570534       1 memcache.go:135] couldn't get resource list for networking.k8s.io/v1beta1: the server could not find the requested resource
E0812 10:59:50.591189       1 memcache.go:135] couldn't get resource list for scheduling.k8s.io/v1: the server could not find the requested resource
E0812 10:59:50.648692       1 memcache.go:135] couldn't get resource list for coordination.k8s.io/v1: the server could not find the requested resource
E0812 10:59:50.748724       1 memcache.go:135] couldn't get resource list for node.k8s.io/v1beta1: the server could not find the requested resource
I0812 11:00:02.290098       1 event.go:221] Event(v1.ObjectReference{Kind:"Node", Namespace:"", Name:"c8r001", UID:"061df7cf-9da0-11e9-833b-005056b0ebb3", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeNotReady' Node c8r001 status is now: NodeNotReady
E0812 11:00:03.979644       1 resource_quota_controller.go:407] unable to retrieve the complete list of server APIs: coordination.k8s.io/v1: the server could not find the requested resource, networking.k8s.io/v1beta1: the server could not find the requested resource, node.k8s.io/v1beta1: the server could not find the requested resource, scheduling.k8s.io/v1: the server could not find the requested resource
W0812 11:00:17.683827       1 garbagecollector.go:647] failed to discover some groups: map[scheduling.k8s.io/v1:the server could not find the requested resource networking.k8s.io/v1beta1:the server could not find the requested resource coordination.k8s.io/v1:the server could not find the requested resource node.k8s.io/v1beta1:the server could not find the requested resource]
E0812 11:00:21.015854       1 memcache.go:135] couldn't get resource list for networking.k8s.io/v1beta1: the server could not find the requested resource
E0812 11:00:21.021852       1 memcache.go:135] couldn't get resource list for scheduling.k8s.io/v1: the server could not find the requested resource
E0812 11:00:21.099570       1 memcache.go:135] couldn't get resource list for coordination.k8s.io/v1: the server could not find the requested resource
E0812 11:00:21.202855       1 memcache.go:135] couldn't get resource list for node.k8s.io/v1beta1: the server could not find the requested resource
E0812 11:00:34.430679       1 resource_quota_controller.go:407] unable to retrieve the complete list of server APIs: coordination.k8s.io/v1: the server could not find the requested resource, networking.k8s.io/v1beta1: the server could not find the requested resource, node.k8s.io/v1beta1: the server could not find the requested resource, scheduling.k8s.io/v1: the server could not find the requested resource
W0812 11:00:50.084892       1 garbagecollector.go:647] failed to discover some groups: map[networking.k8s.io/v1beta1:the server could not find the requested resource scheduling.k8s.io/v1:the server could not find the requested resource node.k8s.io/v1beta1:the server could not find the requested resource coordination.k8s.io/v1:the server could not find the requested resource]
E0812 11:00:51.465165       1 memcache.go:135] couldn't get resource list for networking.k8s.io/v1beta1: the server could not find the requested resource
E0812 11:00:51.470694       1 memcache.go:135] couldn't get resource list for scheduling.k8s.io/v1: the server could not find the requested resource
E0812 11:00:51.550410       1 memcache.go:135] couldn't get resource list for coordination.k8s.io/v1: the server could not find the requested resource
E0812 11:00:51.650411       1 memcache.go:135] couldn't get resource list for node.k8s.io/v1beta1: the server could not find the requested resource
I0812 11:00:52.405543       1 event.go:221] Event(v1.ObjectReference{Kind:"Node", Namespace:"", Name:"c8r001", UID:"061df7cf-9da0-11e9-833b-005056b0ebb3", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeNotReady' Node c8r001 status is now: NodeNotReady
E0812 11:01:04.881708       1 resource_quota_controller.go:407] unable to retrieve the complete list of server APIs: coordination.k8s.io/v1: the server could not find the requested resource, networking.k8s.io/v1beta1: the server could not find the requested resource, node.k8s.io/v1beta1: the server could not find the requested resource, scheduling.k8s.io/v1: the server could not find the requested resource

```

**What you expected to happen**:
Upgraded nodes remain in the ready state.

**How to reproduce it (as minimally and precisely as possible)**:
Upgrade from 1.13.7 to 1.14.4?

**Anything else we need to know?**:

Etcd recently upgraded from 3.2.24 to 3.3.13

[root@5n001 k8supgrade]# kubectl version
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.7", GitCommit:"4683545293d792934a7a7e12f2cc47d20b2dd01b", GitTreeState:"clean", BuildDate:"2019-06-06T01:46:52Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.4", GitCommit:"a87e9a978f65a8303aa9467537aa59c18122cbf9", GitTreeState:"clean", BuildDate:"2019-07-08T08:43:10Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}

NAME="Container Linux by CoreOS"
ID=coreos
VERSION=2135.5.0
VERSION_ID=2135.5.0
BUILD_ID=2019-07-01-1959
PRETTY_NAME="Container Linux by CoreOS 2135.5.0 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"

Network plugin and version (if this is a network-related bug): Calico 3.7.4

Coredns 1.5.0

/sig support 
/sig node

**Comment by mattjmcnaughton** [comment]
Hi @cann0nf0dder !

Is the information here helpful at all: https://kubernetes.io/docs/setup/release/version-skew-policy/?

I'm not positive the exact order in which you upgraded your nodes, but I'm wondering if some of the version skew policies may have been violated?

**Comment by cann0nf0dder** [comment]
thanks @mattjmcnaughton 
We have used the same process upgrading from 1.10 -> 1.11 -> 1.12 -1.13 ->1.14 and this is the first time i see this behaviour. 

The process is quite simple:
We update the image on the repo and do a rolling upgrade through each nodes, starting with masters:

```
sudo sed -i s/1.13.7/1.14.4/g /etc/kubernetes/kubelet.env >/dev/null
sudo sed -i s/hyperkube:$1.13.7/hyperkube:$1.14.4/g /etc/kubernetes/manifests/* >/dev/null

sudo systemctl daemon-reload && sudo systemctl restart kubelet
```
Currently all 4 masters are on 1.13.7 and the one flapping on 1.14
I may add, I upgraded around 10 clusters of which all had the flapping in/out NotReady without the impact, I'm guessing it's fine since one controller was always ready and when that was done and all  are upgraded to 1.14, they remain Ready and happy. I just wanted to dig into it more as it bothers me :D 

Not sure if there are any other particular logs I could look into...

**Comment by liggitt** [comment]
possibly related to https://github.com/kubernetes/enhancements/pull/1201 (controller talking to older API server which is serving discovery docs advertising API groups it doesn't actually have, and which fail discovery)

**Comment by cann0nf0dder** [comment]
@liggitt 
Intersting observation:
The upgraded master (1.14) would be flapping to NodeReady if the controller-manager leader is running on any other master still on previous version (1.13). Once I rebooted master which had leader on it and the upgraded controller was elected to be the leader, the issue no longer exists.

Is there any way I can trigger leader election without finding out who the leader is and kicking it?

**Comment by cann0nf0dder** [comment]
hmm in fact re-election itself doesn't make c8r01 leader when i tried again, leader is still elected at random, so looking for other ways to set newly upgraded master as leader...

**Comment by liggitt** [comment]
to be clear, is it the controller-manager process that is restarting, or the apiserver as well?

**Comment by cann0nf0dder** [comment]
> to be clear, is it the controller-manager process that is restarting, or the apiserver as well?

not sure if I follow. I'm still unable to track what exactly triggers the upgraded master to keep flapping in/out node ready. (most of what I found is in the logs above) What I found is, if the elected (leader) master runs on upgraded version (1.14.4), the issue is not present. Since I have 5 masters, I'm looking for a process to upgrade 1st master and make that a leader, so far the election seems to be random, more often than not a master on 1.13.7 assumes the leader status.

**Comment by cann0nf0dder** [comment]
To be specific I meant kube-control-manager on the upgraded master needs to be a leader for the issue to resolve.

**Comment by matthias50** [comment]
We were experiencing the same issue.  Passing --feature-gates=NodeLease=false to kubelet caused the flapping to stop when kubelet is 1.14 and masters are not yet updated to 1.14. So our solution is to turn NodeLease off for kubelet (and only for kubelet) till the masters are updated to 1.14 and then remove the feature gate override in a second update to the cluster.

**Comment by cann0nf0dder** [comment]
Thanks for the info @matthias50 . I've also just came across release notes for 1.14.5 and it looks to have been addressed and fixed by #80173 . I wish i was upgrading to 1.14.5 and not 1.14.4 - it took a while to try to figure this one out! This can be closed now, thanks all for chipping in!

**Comment by bcorijn** [comment]
Out of completeness, I ran into this issue while updating my cluster from 1.13.7 to 1.14.8 with kops 1.14. So it didn't look like the fix in https://github.com/kubernetes/kubernetes/pull/80173 actually solved this for me (or other kops users in Slack).
As soon as the last master got stopped (so it didn't even need to come back online) the others stabilized.

**Comment by wskulley** [comment]
> Out of completeness, I ran into this issue while updating my cluster from 1.13.7 to 1.14.8 with kops 1.14. So it didn't look like the fix in #80173 actually solved this for me (or other kops users in Slack).
> As soon as the last master got stopped (so it didn't even need to come back online) the others stabilized.
> 
> The nodes don't report any issue, they actually only report healthy logs:
> 
> ```
> kubelet[4995]: I1017 16:49:35.070164    4995 kubelet_node_status.go:468] Recording NodeReady event message for node ip-10-20-71-14.ec2.internal
> kubelet[4995]: I1017 16:49:35.070139    4995 kubelet_node_status.go:468] Recording NodeHasSufficientPID event message for node ip-10-20-71-14.ec2.internal
> kubelet[4995]: I1017 16:49:35.070120    4995 kubelet_node_status.go:468] Recording NodeHasNoDiskPressure event message for node ip-10-20-71-14.ec2.internal
> kubelet[4995]: I1017 16:49:35.070078    4995 kubelet_node_status.go:468] Recording NodeHasSufficientMemory event message for node ip-10-20-71-14.ec2.internal
> kubelet[4995]: I1017 16:48:45.039052    4995 kubelet_node_status.go:468] Recording NodeReady event message for node ip-10-20-71-14.ec2.internal
> kubelet[4995]: I1017 16:48:45.038778    4995 kubelet_node_status.go:468] Recording NodeHasSufficientPID event message for node ip-10-20-71-14.ec2.internal
> kubelet[4995]: I1017 16:48:45.038759    4995 kubelet_node_status.go:468] Recording NodeHasNoDiskPressure event message for node ip-10-20-71-14.ec2.internal
> kubelet[4995]: I1017 16:48:45.038717    4995 kubelet_node_status.go:468] Recording NodeHasSufficientMemory event message for node ip-10-20-71-14.ec2.internal
> ```

I observed same 1.13.10 and .11 -> 1.14.6
